[GStreamer][Holepunch] Do not initialize GStreamer when using external holepunch

hwti opened the pull request:
#### dffc4f0ce080b06153c44e2e3ba5e9534df7bdda
<pre>
[GStreamer][Holepunch] Do not initialize GStreamer when using external holepunch
<a href="https://bugs.webkit.org/show_bug.cgi?id=267532">https://bugs.webkit.org/show_bug.cgi?id=267532</a>

Reviewed by Philippe Normand.

When both USE(GSTREAMER) and USE(EXTERNAL_HOLEPUNCH) are enabled, when using the external holepunch,
bestMediaEngineForSupportParameters calls MediaPlayerPrivateGStreamer::supportsType, which
initializes GStreamer and the registry scanner just to tell that video/holepunch is not supported.

* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::supportsType):
* Source/WebCore/platform/graphics/holepunch/MediaPlayerPrivateHolePunch.h:

Canonical link: <a href="https://commits.webkit.org/273057@main">https://commits.webkit.org/273057@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a4d95ed2d38c42eb1604546816a49fd21b1d5ee5

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/34040 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/12834 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/36010 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/36677 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/30852 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/35112 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/15218 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/9974 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/29899 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/34541 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/10857 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/30377 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/9468 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/9580 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/30397 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/37985 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/30924 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/30720 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/35682 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/9706 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/7633 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/33585 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/11492 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7857 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/10271 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/10517 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

philn commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
```diff
@@ -2831,17 +2835,25 @@ MediaPlayer::SupportsType MediaPlayerPrivateGStreamer::supportsType(const MediaE
         return result;
 
+#if USE(EXTERNAL_HOLEPUNCH)
+    HashSet<String> externalHolePunchTypes;
+    MediaPlayerPrivateHolePunch::getSupportedTypes(externalHolePunchTypes);
```
Can this be done `once` only?

hwti replied:
`MediaPlayerPrivateHolePunch` already has a static `HashSet`, so there would be no point adding another static variable here.
If we want to avoid the copy done by `getSupportedTypes()`, we could make `mimeTypeCache()` a member (maybe with another name) to directly use the reference.

philn replied:
It's ok as it is i think. Thanks

hwti committed:
[GStreamer][Holepunch] Do not initialize GStreamer when using external holepunch
https://bugs.webkit.org/show_bug.cgi?id=267532

Reviewed by Philippe Normand.

When both USE(GSTREAMER) and USE(EXTERNAL_HOLEPUNCH) are enabled, when using the external holepunch,
bestMediaEngineForSupportParameters calls MediaPlayerPrivateGStreamer::supportsType, which
initializes GStreamer and the registry scanner just to tell that video/holepunch is not supported.

* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::supportsType):
* Source/WebCore/platform/graphics/holepunch/MediaPlayerPrivateHolePunch.h:

Canonical link: https://commits.webkit.org/273057@main
